src/flight.rs

```diff
@@ -193,3 +202,204 @@ impl SqlFlightClient {
         Ok(stream)
     }
 }
+
+pub fn query_to_stream_with_params(
+    client: Arc<SqlFlightClient>,
+    sql: &str,
+    params: Option<RecordBatch>,
+) -> RetryableQueryStream {
+    RetryableQueryStream::new(client, sql, params)
+}
+
+/// Represents the current state of the `RetryableQueryStream` state machine.
+///
+/// The stream transitions through these states during query execution:
+/// `Ready` → `Executing` → `Streaming` → `Error` (`is_retryable`: True) → `Ready`
+/// or
+/// `Ready` → `Executing` → `Error` (`is_retryable`: True) → `Ready`
+/// or
+/// `Ready` → `Executing` → `Error` (`is_retryable`: False) -> `Terminated`.
+enum StreamState {
+    /// Initial state, ready to start or retry a query
+    Ready,
+    /// Query is being executed, waiting for the server to return a stream
+    Executing(Pin<Box<dyn Future<Output = Result<FlightRecordBatchStream, GenericError>> + Send>>),
```

Comment:
Related to my previous comment, this state would still exist but would only be entered when a retry is performed mid-stream. The outer function that initiates the stream would retry until it gets a valid `FlightRecordBatchStream` and then it could pass it into a new constructor for this Stream that automatically puts it into the `Streaming` state. And then if a mid-stream error occurs, then the stream would reset from `Streaming` -> `Error` -> `Ready` -> `Executing` -> `Streaming`